cpp/src/arrow/compute/kernels/scalar_cast_string.cc

```diff
@@ -58,7 +58,7 @@ struct NumericToStringCastFunctor {
 
   static Status Convert(KernelContext* ctx, const ArrayData& input, ArrayData* output) {
     FormatterType formatter(input.type);
-    BuilderType builder(input.type, ctx->memory_pool());
+    BuilderType builder(utf8(), ctx->memory_pool());
```

Comment:
IIUC, this will cause the output to be typed `utf8` even if we're casting to `large_utf8`. I think it's better to rely on the default constructor (which sets the correct output type without the need to be explicit):

```suggestion
    BuilderType builder(ctx->memory_pool());
```

Reply:
Done